amxmodx/CLibrarySys.cpp

```diff
@@ -0,0 +1,559 @@
+// vim: set ts=4 sw=4 tw=99 noet:
+//
+// AMX Mod X, based on AMX Mod by Aleksander Naszko ("OLO").
+// Copyright (C) The AMX Mod X Development Team.
+//
+// This software is licensed under the GNU General Public License, version 3 or higher.
+// Additional exceptions apply. For full license details, see LICENSE.txt or visit:
+//     https://alliedmods.net/amxmodx-license
+
+#include "CLibrarySys.h"
+#include <amxmodx.h>
+
+LibrarySystem g_LibSys;
+
+/******************/
+/* Directory Code */
+/******************/
+
+CDirectory::CDirectory(const char *path)
+{
+#if defined PLATFORM_WINDOWS
+
+	char newpath[PLATFORM_MAX_PATH];
+	UTIL_Format(newpath, sizeof(newpath) - 1, "%s\\*.*", path);
+
+	m_dir = FindFirstFile(newpath, &m_fd);
+
+	if (!IsValid())
+	{
+		m_fd.cFileName[0] = '\0';
+	}
+
+#elif defined PLATFORM_POSIX
+
+	m_dir = opendir(path);
+
+	if (IsValid())
+	{
+		m_ep = readdir(m_dir); // TODO: we need to read past "." and ".."!
+		UTIL_Format(m_origpath, sizeof(m_origpath) - 1, "%s", path);
+	}
+	else
+	{
+		m_ep = nullptr;
+	}
+#endif
+}
+
+CDirectory::~CDirectory()
+{
+	if (IsValid())
+	{
+#if defined PLATFORM_WINDOWS
+
+		FindClose(m_dir);
+
+#elif defined PLATFORM_POSIX
+
+		closedir(m_dir);
+#endif
+	}
+}
+
+void CDirectory::NextEntry(cell* offset)
+{
+#if defined PLATFORM_WINDOWS
+
+	if (offset)
+	{
+		// Should be declared after loop so entry starts to '.' and not '..'
+		// But old code did that, so keep this for compatibility.
+		++*offset;
```

Comment:
I think keeping this function (and the class in general) free from `read_dir` quirks is the better way to go. `CDirectory` should cleanly provide access to a directory, so making legacy quirks part of it's interface is bad.

The offset stuff in general should probably not be part of `NextEntry` at all. It should only do what it says, so it can be used for a potential SM-style alternative to `read_dir` in the future. If we want to keep the seekdir/telldir optimization for POSIX we should move the offset stuff out into another function like `MoveEntry`, `SkipEntry` or similar.